hal/src/stm32f2xx/usb_hal.c

```diff
@@ -148,9 +155,14 @@ void USB_USART_LineCoding_BitRate_Handler(void (*handler)(uint32_t bitRate))
  *******************************************************************************/
 uint8_t USB_USART_Available_Data(void)
 {
-    if(USB_Rx_State == 1)
+    if (USB_USART_Connected())
     {
-        return (USB_Rx_length - USB_Rx_ptr);
+        int32_t available = 0;
```

Comment:
I'm concerned that USB_Rx_Buffer_head / USB_Rx_Buffer_tail will be changed by the USB IRQ after being tested. 

As a minimum, capture the tail and head into a local variables, and use that so we are certain they are consistent.

A context switch after comparing head >= tail could see head roll over to the start of the buffer, which would then result in the incorrect size being returned.